ethereum/geth_client.go

```diff
@@ -59,217 +58,251 @@ type Config interface {
 
-		err = gethClient.incrementNonce(opts)
+		current++
+		err = incrementNonce(opts, gc.config.GetTxPoolAccessEnabled(), gc.client, gc.rpcClient)
 		if err != nil {
-			return
+			return nil, fmt.Errorf("failed to increment nonce: %v", err)
+		}
+
+		if opts.Nonce != nil {
+			log.Infof("Incrementing Nonce to [%v]\n", opts.Nonce.String())
+		}
+
+		var in []reflect.Value
+		in = append(in, reflect.ValueOf(opts))
+		for _, p := range params {
+			in = append(in, reflect.ValueOf(p))
 		}
-		in = make([]reflect.Value, len(params)+1)
-		val := reflect.ValueOf(opts)
-		in[0] = val
-		for k, param := range params {
-			in[k+1] = reflect.ValueOf(param)
+
+		result := f.Call(in)
+		var tx *types.Transaction
+		if !result[0].IsNil() {
+			tx = result[0].Interface().(*types.Transaction)
 		}
-		result = f.Call(in)
-		tx = result[0].Interface().(*types.Transaction)
-		err = nil
-		if result[1].Interface() != nil {
+
+		if !result[1].IsNil() {
 			err = result[1].Interface().(error)
 		}
 
-		if err != nil {
-			if (err.Error() == TransactionUnderpriced) || (err.Error() == NonceTooLow) {
-				log.Warningf("Concurrent transaction identified, trying again [%d/%d]\n", current, maxTries)
-				time.Sleep(gethClient.config.GetEthereumIntervalRetry())
-			} else {
-				done = true
-			}
-		} else {
-			done = true
+		if err == nil {
+			return tx, nil
 		}
+
+		if (err.Error() == transactionUnderpriced) || (err.Error() == nonceTooLow) {
+			log.Warningf("Concurrent transaction identified, trying again [%d/%d]\n", current, maxTries)
+			time.Sleep(gc.config.GetEthereumIntervalRetry())
+			continue
+		}
+
+		return nil, err
 	}
 
-	return
 }
 
-func (gethClient *GethClient) incrementNonce(opts *bind.TransactOpts) (err error) {
-	if !gethClient.config.GetTxPoolAccessEnabled() {
-		log.Warningf("Ethereum Client doesn't support txpool API, may cause concurrency issues.")
-		return
+// noncer defines functions to get the next nonce
+type noncer interface {
+	PendingNonceAt(ctx context.Context, account common.Address) (uint64, error)
+}
+
+// callContexter defines functions to get CallContext
+type callContexter interface {
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
+// incrementNonce updates the opts.Nonce to next valid nonce
+// We pick the current nonce by getting latest transactions included in the blocks including pending blocks
+// then we check the txpool to see if there any new transactions from the address that are not included in any block
+// If there are no pending transactions in txpool, we use the current nonce + 1
+// else we increment the greater of current nonce or the nonce derived from txpool
+func incrementNonce(opts *bind.TransactOpts, txpoolAccessEnabled bool, noncer noncer, cc callContexter) error {
+	// check if the txpool access is enabled
+	if !txpoolAccessEnabled {
+		log.Warningf("Ethereum Client doesn't support txpool API, may cause transactions failures")
+		opts.Nonce = nil
```

Comment:
Why do we need to nil nonce here? is it what is passed by default right?

Reply:
I was passed as nil by default but we were `nil`ing nonce at GetOpts. Since this `incrementNonce` is the only place we check the nonce, I felt this is the right place. 

Coming to `nil`ing the nonce here, since the `txnpoolaccess` is not enabled, we will just let the geth-node pick the nonce from the pending transactions.

Reply:
We still need to nil the `opts.nonce` at the beginning of any new tx request otherwise we will have a problem with high concurrency as cached opts per account will be outdated.

Reply:
Yes, this will be the place it gets either nil if the `transaction pool access` is not enabled. If enabled, we will get the nonce from the blocks including pending and then we will get the txns from the pool, get the highest nonce and and increment it ourselves.

Reply:
ok, now looks good as we default to the pending nonce a bit below. Thanks!